### PR TITLE
Improve debian systemd service enable/start logic in post installation scripts

### PR DESCRIPTION
**Pull request by reubenmiller**
## Proposed changes

The following changes are covered by this PR:
* Remove conditional logic in postinst commands (where possible), as the `#DEBHELPER#` was already starting and enabling the service, so the additional call to start and enable the service was unnecessary in all services except (`tedge-mapper` and `tedge-agent`)
* Avoid using `systemctl` directly and use the systemd debian helpers, `deb-systemd-invoke` and `deb-systemd-helper`. The helper commands respect the [policy-rc.d](https://manpages.ubuntu.com/manpages/focal/en/man8/policy-rc.d-declarative.8.html) which gives the user more control over the services, for example the user can decide if they should be installed or not.

**Advantages**

* Installing the debian packages does not fail due to a call to `systemctl start <service>` because systemd is not running as PID 1 during the docker build stage. This was previously stopping the `tedge-watchdog` from being installed when building a docker image.
* Allow users to control whether services are installed or not via `policy-rc.d`

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

References

* [deb-systemd-invoke](https://www.commandlinux.com/man-page/man1/deb-systemd-invoke.1p.html)
* [policy-rc.d](https://manpages.ubuntu.com/manpages/focal/en/man8/policy-rc.d-declarative.8.html)

### CODE REVIEW

**Review by rina23q — COMMENTED** [review]
Comments 1 of 3
---

By recalling old days,

1. `tedge-agent` and `tedge-mapper` services will start/stop and be enabled/disabled by `tedge connect` command. Thus, making `tedge-mapper` and `tedge-agent` have/used to have these configuration in `Cargo.toml` file.
```toml
[package.metadata.deb.systemd-units]
enable = false
start = false
stop-on-upgrade = false
```

2. Then, the condition to c8y-bridge file was introduced. If user already connected to c8y, the mapper and agent must be started.

3. Next, we got a wrong assumption and made a mistake. All other services (e.g. `tedge-watchdog`, `c8y-log-plugin` etc.) won't get started/enabled by default, which is wrong. So, applied the same condition to the c8y-bridge file.

In short,
- Removing start/enable block from postinst of `tedge-watchdog` is totally correct. It is useless as already DEBHELPER does the same.
- All other services except `tedge-agent` and `tedge-mapper`, it is also correct, because already DEBHELPER enables/starts the service regardless of bridge file existence.

**Comment by PradeepKiruvale** [comment]
> By recalling old days,
> 
> 1. `tedge-agent` and `tedge-mapper` services will start/stop and be enabled/disabled by `tedge connect` command. Thus, making `tedge-mapper` and `tedge-agent` have/used to have these configuration in `Cargo.toml` file.
> 
> ```toml
> [package.metadata.deb.systemd-units]
> enable = false
> start = false
> stop-on-upgrade = false
> ```
> 
> 2. Then, the condition to c8y-bridge file was introduced. If user already connected to c8y, the mapper and agent must be started.
> 3. Next, we got a wrong assumption and made a mistake. All other services (e.g. `tedge-watchdog`, `c8y-log-plugin` etc.) won't get started/enabled by default, which is wrong. So, applied the same condition to the c8y-bridge file.
> 
> In short,
> 
> * Removing start/enable block from postinst of `tedge-watchdog` is totally correct. It is useless as already DEBHELPER does the same.
> * All other services except `tedge-agent` and `tedge-mapper`, it is also correct, because already DEBHELPER enables/starts the service regardless of bridge file existence.

But what is the point in starting the ` c8y log/config` services without a bridge? They will fail to do anything. Isn't it better to start them only if it's connected to c8y?

**Review by rina23q — COMMENTED** [review]
Comments 2 of 3
---

Now need to think about `tedge-agent` and `tedge-mapper`. They are special because of self-update by c8y.

[tedge-agent]
As we stated in `Cargo.toml`, we cannot use DEBHELPER because for self-update,
1. We must keep `tedge-agent` service running with the old binary, otherwise, software update will terminate in the middle of processing the package upgrade.
2. User needs to restart the service manually somehow to apply the new binary.

The current conditions below means
a. `[ -f "/etc/tedge/mosquitto-conf/c8y-bridge.conf" ]` => tedge device is connected to c8y
b. `! systemctl is-active --quiet tedge-agent` => But `tedge-agent` is not running. This means that it is not doing self-update.
In short, so it expects the situation where user installs the `tedge-agent` newly after they connected to c8y.

**Review by rina23q — COMMENTED** [review]
Comments 3 of 3
---

[tedge-mapper]
`DEBHELPER` doesn't start/enable services due to the configuration on `Cargo.toml`.
By design, all mapper services besides `tedge-mapper-collectd` are supposed to be started and enabled by `tedge connect <cloud>`. That's why we can see the conditions for bridge files.

---
Long story short, I really think what you made in this PR makes sense.

**Comment by reubenmiller** [comment]
> 
> But what is the point in starting the ` c8y log/config` services without a bridge? They will fail to do anything. Isn't it better to start them only if it's connected to c8y?
> 
> If started without bridge then it will be keep printing the below logs, for example `log-plugin`
> 
> ```
> 2023-03-29T12:13:07.655198191Z ERROR init: c8y_api::http_proxy: An error occurred while retrieving internal Id, operation will retry in 60 seconds and mapper will reinitialise.
> 
>  Error: RequestTimeout
> ```

Though the plugins were already starting without the bridge settings anyway as the DEBHELPER logic was running before the if statement. However moving forward we need to be careful about adding dependencies at install time, and ideally we need to remove the dependencies between the services. If the user does not want the `c8y-configuration-plugin` then they are free to uninstall it, or disable it themselves (as the disabling will be sticky).

In general we need to shift our thinking that all services have access to the same file system, because when deployed in single-process containers, each process might not have access to the same shared filesystem. But conceptually, the c8y-configuration-plugin does not care that the `c8y-bridge.conf` file exists, it really wants to know "am I going to receive any messages?". We should be making this state information available via an MQTT topic and not via the filesystem.

**Comment by PradeepKiruvale** [comment]
> > But what is the point in starting the ` c8y log/config` services without a bridge? They will fail to do anything. Isn't it better to start them only if it's connected to c8y?
> > If started without bridge then it will be keep printing the below logs, for example `log-plugin`
> > ```
> > 2023-03-29T12:13:07.655198191Z ERROR init: c8y_api::http_proxy: An error occurred while retrieving internal Id, operation will retry in 60 seconds and mapper will reinitialise.
> > 
> >  Error: RequestTimeout
> > ```
> 
> Though the plugins were already starting without the bridge settings anyway as the DEBHELPER logic was running before the if statement. However moving forward we need to be careful about adding dependencies at install time, and ideally we need to remove the dependencies between the services. If the user does not want the `c8y-configuration-plugin` then they are free to uninstall it, or disable it themselves (as the disabling will be sticky).
> 
I agree we must get rid of these dependencies between the services.

> In general we need to shift our thinking that all services have access to the same file system, because when deployed in single-process containers, each process might not have access to the same shared filesystem. But conceptually, the c8y-configuration-plugin does not care that the `c8y-bridge.conf` file exists, it really wants to know "am I going to receive any messages?". We should be making this state information available via an MQTT topic and not via the filesystem.

Yes, we can check if the bridge is up or not using the `bridge health status` message. But that can be only for trying to loop around before it requests the `jwt` token to get the `internal-id`. 
If we do that, still it will be looping around and it consumes resources since we are talking about edge devices, I am a bit concerned about the resources and power consumption as well. I do not know how much priority we give to such things here as of now.

**Comment by reubenmiller** [comment]
As of thin-edge.io 0.10.0, the behaviour of services after installation is as follows:

|Service|Started|Enabled|
|--|--|--|
|tedge-agent|no|no|
|tedge-mapper-c8y|no|no|
|tedge-mapper-az|no|no|
|tedge-mapper-aws|no|no|
|tedge-mapper-collectd|yes|no|
|tedge-watchdog|yes|yes|
|c8y-log-plugin|yes|yes|
|c8y-configuration-plugin|yes|yes|
|c8y-firmware-plugin|yes|yes|

**Notes**

* `tedge-mapper-collect` is enabled if collectd service is running, but the service is not enabled (meaning it will not be started after a device reboot)
* `tedge-mapper-<c8y|az|aws>` services are only started and enabled if the corresponding bridge configuration is set

**After**

The proposed PR changes would only change the `tedge-mapper-collectd` service, where it would be enabled now (if the `collectd` service is running (which is the more logical behaviour anyway, as why start a service without enabling it?).

|Service|Started|Enabled|
|--|--|--|
|tedge-agent|no|no|
|tedge-mapper-c8y|no|no|
|tedge-mapper-az|no|no|
|tedge-mapper-aws|no|no|
|tedge-mapper-collectd|yes|yes|
|tedge-watchdog|yes|yes|
|c8y-log-plugin|yes|yes|
|c8y-configuration-plugin|yes|yes|
|c8y-firmware-plugin|yes|yes|

**Comment by reubenmiller** [comment]
@PradeepKiruvale Resource usage associated to unnecessary services running is not something addressed by this PR. The PR only looks to address the inconsistent and duplicated logic which is currently in the maintainer scripts.

However if a device maintainer can always control which services are run and which are not on their own, there is nothing stopping them from implementing their own logic on top of the services.

**Comment by didier-wenzek** [comment]
> ideally we need to remove the dependencies between the services.

Yes, definitely. The thin-edge daemons should be able to start in any order without strong dependencies. A service might be in some degraded mode till another another service is up but it should not stop or crash. Unfortunately this is not what we have today as pointed out by the fact we have so often to run `tedge reconnect c8y`.

**Review by rina23q — APPROVED** [review]

**Review by PradeepKiruvale — APPROVED** [review]
